diff.py

```diff
@@ -1824,21 +1884,9 @@ def process(dump: str, config: Config) -> List[Line]:
         # transforming 'row' into a coarser version that ignores registers and
         # immediates.
         original = row
-
-        while i < len(lines):
-            reloc_row = lines[i]
-            if "R_AARCH64_" in reloc_row:
-                # TODO: handle relocation
-                pass
-            elif "R_MIPS_" in reloc_row:
-                original = process_mips_reloc(reloc_row, original, arch)
-            elif "R_PPC_" in reloc_row:
-                original = process_ppc_reloc(reloc_row, original)
-            elif "R_ARM_" in reloc_row:
-                original = process_arm_reloc(reloc_row, original, arch)
-            else:
-                break
-            i += 1
+        if re.search(arch.re_reloc, row):
+            processed_row = reloc_processor.process(row, output)
+            original = processed_row if processed_row else row
```

Comment:
I'm not too happy about making this a post-processing step that fixes up the previous row. We had it like that before, and it resulted in confusion and bugs. For instance you have the following bugs with this version of the code:

- `data_refs` will apply twice: one to the real line, once to the relocation
- `mnemonic` will be set wrongly (because this code runs too late) (but if you move it up much higher the hexify_int step will happen after relocation processing which I'm not sure if it has side effects)
- `comment`s get stripped I think
- `lui $v0, %hi(var)`, `jr $ra`, `addiu $v0, $v0, %lo(var)` won't process relocations for the last line with MIPS and --stop-jr-ra.

Perhaps more importantly the current way *feels* more right to me: the relocations are processed together with the current instruction as part of parsing it.

Using a separate class for processing relocations is a nice idea. I think you can keep that but change it to use the previous system? Would be fine to merge with DifferenceNormalizer too I think. ARM32_LOAD_POOL_PATTERN will need special treatment but that's fine: it isn't a relocation, it's something else.

Reply:
I kept the relocation classes and moved the logic marking data pool annotations into DifferenceNormalizer. I added a PostProcessor class -- for ARM32, it adds symbols in the final annotation by checking all the output lines and dereferencing addresses.